### PR TITLE
[Audit] [SC-713] Feat: Add DepositVault

### PR DESCRIPTION
**Pull request by FHieser**

### CODE REVIEW

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]: Not being used

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]:
```suggestion
 * @notice  This contract allows users to deposit tokens to fund the workflow.
```

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]:
```suggestion
    /// @dev Storage gap for future upgrades
```

Otherwise this won't be picked up by the automatic doc generator from Forge. This hasn't been done in the other contracts yet.

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]: Not being used

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]: Not that important, but we could maybe move this modifier to the Module_v1, as it is being used in both the BondingCurveBase, and prob other modules as well

**Review by Zitzak — CHANGES_REQUESTED** [review]
Few small thinks, nothing major.

We are of course bit short on time so not made a comment of it, but would be great if the test functions can be written as defined by the testing handbook :) So not a requirement as of now, but we should start keeping ourselves to it

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (FHieser) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]: Should I do that here?

Comment (Zitzak) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol]: Not sure. Just noticed that the same functionality is needed across multiple modules. I'll leave it up to you :)

**Review by Zitzak — APPROVED** [review]
LGTM

**Comment by Zitzak** [comment]
@FHieser @0xNuggan I decided to merge the FM already so we can sent a link to the auditors today. Jfyi

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:95]: To be consistent with our other FundingManagers and what's common in the space, we should have a `depositFor(address, amount)` function here imho.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol]: The `Note` should have it's only `@dev` tag according to the best practices of NatSpec, as it is quite important.

File: src/modules/fundingManager/rebasing/interfaces/IRebasingERC20.sol
Comment (marvinkruse) [inline src/modules/fundingManager/rebasing/interfaces/IRebasingERC20.sol]: Description of the `error` is wrong.

File: src/modules/fundingManager/rebasing/interfaces/IRebasingERC20.sol
Comment (marvinkruse) [inline src/modules/fundingManager/rebasing/interfaces/IRebasingERC20.sol]: Description of the `error` is slightly wrong (_the problem is not that there is a cap, the problem is that it was reached_).

**Review by marvinkruse — CHANGES_REQUESTED** [review]

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (Zitzak) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:95]: btw, just noticing this function in the FM_Rebasing_v1
https://github.com/InverterNetwork/contracts/blob/dev/src/modules/fundingManager/rebasing/FM_Rebasing_v1.sol#L150

Comment (0xNuggan) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:95]: Do we really need it in this case? The FundingManager does not track balances, so you cannot deposit "for" somebody. In the other cases the depositFor/buyFor takes funds from the sender and increases the balance of somebody else, but here the implementation of a buyFor function would look exactly the same...

**Comment by 0xNuggan** [comment]
took the liberty to update the comments following @marvinkruse  review.

**Review by 0xNuggan — APPROVED** [review]

**Comment by Zitzak** [comment]
@0xNuggan Please don't do more :D was just about to push everything

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:95]: > The FundingManager does not track balances, so you cannot deposit "for" somebody.

I see your point, but to me this doesn't make a big difference. A `depositFor` should still exist, as we do not know whether someone might need it (there are instances where multisigs just give an approval and a specific member can handle the deposit up to the allowance limit whenever they see fit from their own wallet), so I'd still add it just in case, personally.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol]: Formatting is off now, after the change.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol]: Closing this, as it's fixed overall, just the formatting is now off, but I added a new comment for that.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (Zitzak) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol]: I'll fix it. Pushing in a min, after removing my comment edits

**Review by marvinkruse — APPROVED** [review]

**Comment by Zitzak** [comment]
This PR has been send to Omega. Please do not merge until after their review jfyi.
@FHieser @marvinkruse @0xNuggan

**Comment by marvinkruse** [comment]
(I will rebase this one)

**Comment by marvinkruse** [comment]
@Zitzak @0xNuggan You reviewed it already - but I now rebased this PR after the extensive `dev` changes, only real difference here is now that I activated the prepared lines (that Felix already added) in the new deployment scripts and made sure it all runs fine. If you could check that out briefly it would be great 🙏

**Review by marvinkruse — APPROVED** [review]

**Review by Zitzak — APPROVED** [review]
LGTM

**Review by 0xNuggan — APPROVED** [review]
LGTM

**Comment by marvinkruse** [comment]
Note: This PR will now be merged. We will still monitor this PR as it will be used for a review by Omega. If there should be required fixes, a new PR will be opened up towards `dev`.

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:112]: So if I approved 20 tokens to this contract, but only wanted to deposit 10, anyone can call depositFor and make it deposit the other 10 tokens as well. This seems quite dangerous. A normal depositFor behaviour would be that the tokens are taken from the msgSender, but are credited to the from address.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:25]: The implementation does not revert on self deposit

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:17]: The implementation does not revert on self deposit

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:102]: Missing a check that the from address is not address(this)

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:111]: Should probably disallow also 0 amount here

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:112]: I'd say what is implemented here is more "depositFrom" than the usual "depositFor"

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:14]: Unused import

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:2]: Why is the pragma in some contracts ^0.8.0 but in others 0.8.23

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (ben-kaufman) [inline src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:11]: You might want to add here address indexed _to, to allow an account to deposit into another's "credit"

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:25]: Valid, the `@dev`-comment is a copy-paste-error from `FM_Rebasing_v1`, as self-deposits would be bad in that contract. 
Here, I'd argue that a `depositFor` with `from` being set to the address of the contract itself (`address(this)`) would fail anyway via the approval check (throws `ERC20InsufficientAllowance`), so there is no way for self-deposits to happen. So I'd just remove this comment for the `deposit` and `depositFor` functions, especially that - even if a self-deposit would happen (i.e. if for some reason the vault has set an approval for _itself_ - it wouldn't have any effect besides a confusing event emission.

Would you agree with logic @ben-kaufman?

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:17]: Discussed above, see [here](https://github.com/InverterNetwork/contracts/pull/626#discussion_r1707911078).

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (ben-kaufman) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:25]: Yes it should fail on approve check (unless a "special" token implementations don't check approve if from is msg.sender). I think you can either just remove the comment, or require the from address is not address(this).

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:102]: Also discussed above ([here](https://github.com/InverterNetwork/contracts/pull/626#discussion_r1707911078)). I'd argue that this would just be a waste of gas to check, as a self-deposit can only happen if the contract has for some reason self-approved before, and even then it's just a weird `Deposit` being emitted with a false value, no funds are moved. 

Do you think the check should really be added @ben-kaufman?

Comment (ben-kaufman) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:102]: I mostly commented this because it appeared as a requirement in the documentation, but I don't think it's so necessary

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:111]: Yeah, fair. Not a big fan of wasting gas for every user just to ensure that a potential 0 depositing user doesn't pay gas/send their tx if they forgot to put something in the `amount` field, but I believe we do it in other contracts as well, so I guess that's a fair point.

As a 0-deposit has literally zero impact here though and should just be filtered on the UI/SDK level, I'd personally lean towards not changing anything, unless @Zitzak or @0xNuggan think it makes sense.

Or did I miss any security-related implication you had in mind @ben-kaufman?

Comment (ben-kaufman) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:111]: No that's mostly a matter of prefence, shouldn't matter much

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:2]: Actual contracts are all `0.8.23` (or should be), while interfaces are all `^0.8.0`, so other projects can import our interfaces easier without needing multiple compiler versions potentially. We just following the example of other projects here.

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:14]: Valid, fixed in [this commit](https://github.com/InverterNetwork/contracts/pull/646/commits/8eb10a27e4b749a98bd0f2dde49414cd32244193). PR that addresses all of the fixes is #646. 🙏

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:112]: Very valid point, this is not the traditional `depositFor` where a user would get receipt tokens, withdraw rights, or anything else in return. It's a bit confusing. I'd probably just rename it then to `depositFrom`, but...

I'd personally just remove this function. Looking at it now, I don't see any reason to have it. If someone can't (for whatever reason) call the funding manager directly to deposit (weird wallet that can't call arbitrary contracts, or whatever), they can always just transfer the tokens directly into the DepositVault anyway via a regular token transfer, as it doesn't matter who sent the tokens and no accounting happening.

I've removed it in [this commit](https://github.com/InverterNetwork/contracts/pull/646/commits/ef74cf6e825dc685712d95b1f9e3614da9c0eb20). The PR handling all fixes is #646. Let me know what you think @ben-kaufman - I'll also give @Zitzak and @0xNuggan the chance to comment here as well, maybe I missed something, in which case we can still revert my removal of course.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:11]: Now that I removed `depositFor`, this is not applicable anymore, right?
I've removed it in [this commit](https://github.com/InverterNetwork/contracts/pull/646/commits/ef74cf6e825dc685712d95b1f9e3614da9c0eb20). The PR handling all fixes is #646.

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:102]: Okay, makes sense, thank you 🙏 I removed the comment itself in [this commit](https://github.com/InverterNetwork/contracts/pull/646/commits/a31b5938349b9456298d07d30f7f8b900179ea23), so we won't need/add a check itself here. The PR handling all fixes is #646.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:25]: Verified that it fails on the approve check, and that there is no impact besides the event even if that check wouldn't make it fail. I removed the comment itself in [this commit](https://github.com/InverterNetwork/contracts/pull/646/commits/a31b5938349b9456298d07d30f7f8b900179ea23), so we won't need/add a check itself here. The PR handling all fixes is #646.

File: src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/interfaces/IFM_DepositVault_v1.sol:17]: Comment itself removed from the `deposit` function, `depositFor` is removed altogether, so this one should be done as well.

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (marvinkruse) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:111]: Okay, thank you 🙏 I personally wouldn't change anything then, but I'll wait for @Zitzak or @0xNuggan to chime in briefly. 👍

File: src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol
Comment (ben-kaufman) [reply src/modules/fundingManager/depositVault/FM_DepositVault_v1.sol:112]: Yes I think it makes sense to remove it

**Pull request merged**
